### PR TITLE
GEOMESA-2954 Calculate FSDS filters corresponding to partitions

### PR DESCRIPTION
**Pull request by elahrvivaz**
Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Review by aheyne — APPROVED** [review]

File: docs/user/filesystem/commandline.rst [hunk 1]
```diff
@@ -41,6 +41,39 @@ Argument                 Description
...
+
+======================== =========================================================
+Argument                 Description
+======================== =========================================================
+``-p, --path *``         The filesystem root path used to store data
+``-f, --feature-name *`` The name of the schema
+``-q, --cql``            CQL predicate to determine the partitions to operate on
+``--partitions``         Partitions to operate on, subject to the CQL filter
+``--no-header``          Suppress the column headers in the output
+``--config``             Hadoop configuration properties, in the form k=v
```
Comment (jnh5y) [inline docs/user/filesystem/commandline.rst:58]: Some note; I wonder if it'd be better to say "key=value" instead.

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/PartitionScheme.scala [hunk 2]
```diff
@@ -49,6 +60,34 @@ trait PartitionScheme {
...
+  /**
+   * Get a filter that will cover a partitions, i.e. the filter will return all features
+   * in the given partition and none from other partitions
+   *
+   * @param partition partition to cover
+   * @return filter
+   */
+  def getCoveringFilter(partition: String): Filter = {
+    // default impl to prevent API breakage in partition scheme impls
+    // TODO remove default impl in next major release
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/PartitionScheme.scala:88]: Can you add a ticket for this API clean and tag for 4.0.0?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/DateTimeScheme.scala [hunk 3]
```diff
@@ -44,20 +48,54 @@ case class DateTimeScheme(format: String, stepUnit: ChronoUnit, step: Int, dtg:
...
 
     case _ =>
-      dt => ZonedDateTime.parse(fmt.format(dt), fmt) // fall back to format and re-parse
+      dt => DateParsing.parse(fmt.format(dt), fmt) // fall back to format and re-parse
   }
 
   // TODO This may not be the best way to calculate max depth...
   // especially if we are going to use other separators
-  override val depth: Int = format.count(_ == '/')
+  override val depth: Int = format.count(_ == '/') + 1
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/DateTimeScheme.scala:56]: *blinks* where did the +1 come from?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala [hunk 4]
```diff
@@ -13,17 +13,22 @@ import org.locationtech.geomesa.fs.storage.common.partitions.SpatialScheme.Spati
...
 
+  override def pattern: String = s"$bits-bit-xz2"
+
   override def getPartitionName(feature: SimpleFeature): String = {
     val geometry = feature.getAttribute(geom).asInstanceOf[Geometry]
     val envelope = geometry.getEnvelopeInternal
     xz2.index(envelope.getMinX, envelope.getMinY, envelope.getMaxX, envelope.getMaxY).formatted(format)
   }
 
+  override def getCoveringFilter(partition: String): Filter = throw new NotImplementedError()
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:30]: There's no covering filter for an XZ2 partition?

Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:30]: there is but I didn't figure it out...

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/DateTimeScheme.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/DateTimeScheme.scala:56]: this was incorrect, but i think wasn't ever really used. if you think about say a single yearly partition, your pattern is 'yyyy' - depth in that case would be 1, not 0. maybe there are some corner cases here in how you define your pattern with leading/trailing slashes? but at least for the built-in ones, this works correctly.

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/PartitionScheme.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-api/src/main/scala/org/locationtech/geomesa/fs/storage/api/PartitionScheme.scala:88]: added https://geomesa.atlassian.net/browse/GEOMESA-2966

File: docs/user/filesystem/commandline.rst (hunk 1 above)
Comment (elahrvivaz) [inline docs/user/filesystem/commandline.rst:58]:
```suggestion
``--config``             Hadoop configuration properties, in the form key=value
```

Comment (elahrvivaz) [reply docs/user/filesystem/commandline.rst]: updated

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala (hunk 4 above)
Comment (jnh5y) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:30]: whew... can you add a comment + ticket to cover that?  Maybe some a note in the error?

Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:30]: hmm, what note would you like to see? NotImplemented seems pretty clear ;)
created https://geomesa.atlassian.net/browse/GEOMESA-2967 to track the issue

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala [hunk 5]
```diff
@@ -13,17 +13,22 @@ import org.locationtech.geomesa.fs.storage.common.partitions.SpatialScheme.Spati
...
 
   private val xz2 = XZ2SFC((bits / 2).asInstanceOf[Short])
 
+  override def pattern: String = s"$bits-bit-xz2"
+
   override def getPartitionName(feature: SimpleFeature): String = {
     val geometry = feature.getAttribute(geom).asInstanceOf[Geometry]
     val envelope = geometry.getEnvelopeInternal
     xz2.index(envelope.getMinX, envelope.getMinY, envelope.getMaxX, envelope.getMaxY).formatted(format)
   }
```
Comment (elahrvivaz) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:29]:
```suggestion
// TODO https://geomesa.atlassian.net/browse/GEOMESA-2967
```

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala [hunk 6]
```diff
@@ -13,16 +13,21 @@ import org.locationtech.geomesa.fs.storage.common.partitions.SpatialScheme.Spati
...
+// TODO https://geomesa.atlassian.net/browse/GEOMESA-2967
```
Comment (elahrvivaz) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:29]:
```suggestion

  // TODO https://geomesa.atlassian.net/browse/GEOMESA-2967
```

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala (hunk 4 above)
Comment (jnh5y) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:31]: Can we split out something a message referencing that ticket? *shrugs*

Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/XZ2Scheme.scala:31]:
```suggestion
  override def getCoveringFilter(partition: String): Filter =
    throw new NotImplementedError("https://geomesa.atlassian.net/browse/GEOMESA-2967")
```

**Pull request merged**